content/blog/2019-09-25-type_checker.md

```diff
@@ -0,0 +1,88 @@
++++
+title = "Static Type Checking for Bril"
+extra.author = "Yi Jing, Zhijing Li, Neil Adit"
+extra.bio = """
+[Zhijing Li](https://tissue3.github.io/) is a 2nd year Phd student at ECE department, Cornell. She is interested in compiler optimization and domain-specific languages. \n
+[Yi Jiang](http://www.cs.cornell.edu/~yijiang/) is a 2nd year Phd student in CS dept, Cornell. She is interested in designing compilers for Processing-In-Memory related architectures and also general compiler optimization. \n
+[Neil Adit](http://www.neiladit.com) is (guess what!) also a 2nd year PhD student at CSL, Cornell. He is interested in fun things in compiler research.
+
+"""
+extra.latex = true
++++
+
+
+
+## Goal
+
+The goal of the project was to add a static type checker to find type errors, multiple definitions of variables and undefined ones as well. We also ensure that the branch labels are valid and unique.
+
+
+
+## Design and Implementation
+
+### Design
+
+Bril currently supports 2 types - **int** and **bool** which makes type checking relatively easy. Our type checker is defined such that an arithmetic operation  `a = b+c` would raise an error if either the source or destination arguments are *not* int. Similarly, boolean operations only accept all bool arguments and comparison operations have integer arguments but a boolean destination. These type definitions have been nicely defined [here](https://capra.cs.cornell.edu/bril/langref.html). Currently Bril only supports operations on variables and not constants which implies `a = b+5` is an invalid operation. During type checking, we also ensure that the variable has been defined in the function before. Conversely, we make sure that there isn't a redefinition of the variable. For control flow operation, we ensure that labels are present in the code and are uniquely defined. This allows us to treat label strings as a separate type, invisible to the user. 
```

Comment:
> Similarly, it's not quite clear what you mean by "invalid operation" in this:
> 
> > Currently Bril only supports operations on variables and not constants which implies `a = b+5` is an invalid operation.
> 
> Because it's not even clear how you'd write that down in Bril abstract syntax. In particular, this makes it hard to understand whether "invalid operation" means "this is something our type checker must catch" or "this is something that can't even be expressed, so it's something we don't have to worry about."

I thought this is for detecting an operation like:
`a: int = add b 5`
Do you think `5` will act as an undefined variable (and hence won't be detected as a constant integer)?

Reply:
That's up to you to determine. But 5 clearly won't parse as a variable.